### PR TITLE
update to tokio 0.2 release

### PR DESCRIPTION
**Pull request by bluejekyll**
fixes: #931 
fixes: #939

Awaiting releases for:
- [x] tokio-rustls, see https://github.com/quininer/tokio-rustls/pull/56
- [x] h2, see https://github.com/hyperium/h2/pull/428
- [ ] ~typed-headers, https://github.com/sfackler/typed-headers/pull/8~ (separate issue filed, #943)

### CODE REVIEW

**Comment by bluejekyll** [comment]
FYI, I've been told that we probably won't see an H2 release until next week (2019-12-2). Until then, I think this branch should work with default features.

**Comment by encombhat** [comment]
h2 0.2.0 is released just now

**Comment by o01eg** [comment]
`typed-headers` are also used here and not yet updated to the same `http` and `bytes`: https://github.com/sfackler/typed-headers/pull/8

**Comment by bluejekyll** [comment]
FYI, I just rebased this onto master, and force pushed, so you'll want to sync and reset to `[origin|upstream]/update-tokio-0.2-alpha.2`

File: Cargo.toml [hunk 1]
```diff
@@ -11,3 +11,6 @@ members = ["crates/client",
            "tests/integration-tests",
            "bin",
            "util"]
+
+[patch.crates-io]
+typed-headers = { git = "https://github.com/o01eg/typed-headers", branch = "update-dependencies" }
```
Comment (bluejekyll) [inline Cargo.toml]: We need to remove this override before checking in.

File: tests/integration-tests/tests/server_future_tests.rs [hunk 2]
```diff
@@ -342,7 +336,7 @@ fn server_thread_tls(
         .expect("tcp registration failed");
 
     while server_continue.load(Ordering::Relaxed) {
-        io_loop.block_on(tokio_timer::delay(
+        io_loop.block_on(tokio::time::delay(
```
Comment (bluejekyll) [inline tests/integration-tests/tests/server_future_tests.rs]: this should be `delay_for`

File: crates/native-tls/Cargo.toml [hunk 3]
```diff
@@ -45,16 +45,16 @@ name = "trust_dns_native_tls"
...
-tokio-net = "0.2.0-alpha"
-tokio-tls = "0.3.0-alpha"
+tokio = "0.2.1"
+tokio-tls = "0.3.0"
 # disables default features, i.e. openssl...
 trust-dns-proto = { version = "0.18.0-alpha", path = "../proto", features = ["tokio-compat"], default-features = false }
 
 [dev-dependencies]
-tokio = "0.2.0-alpha"
+# tokio = "0.2.0"
```
Comment (bluejekyll) [inline crates/native-tls/Cargo.toml]: remove this

**Comment by bluejekyll** [comment]
FYI, progress here, I've got the server working again. Working through a few test failures. This should be at least usable shortly.

**Comment by bluejekyll** [comment]
Having trouble with the server as a library, works fine in the `named` binary. Trying to figure out why some integration tests are failing.

**Comment by fafhrd91** [comment]
you should remove `tokio` feature as it doesnt make sense anymore

**Comment by bluejekyll** [comment]
That feature will make more sense down the line as we start supporting different executors. As of this moment it isn’t useful, but it will be.

Is it causing you issues in this branch?

**Comment by fafhrd91** [comment]
it is fine. you just need to mark `extern crate tokio` with feature or remove feature. atm resolver does not compile with `default-features = false`

**Comment by bluejekyll** [comment]
Ah. Yes, I’m still cleaning up this branch. I Need to figure out the issue with the integration tests, but I can get that building with default-features off, though you’ll need to specifiy “tokio”

File: crates/resolver/Cargo.toml [hunk 4]
```diff
@@ -56,35 +56,33 @@ serde-config = ["serde", "trust-dns-proto/serde-config"]
...
 
+tokio-compat = ["tokio", "trust-dns-proto/tokio-compat"]
+
 [lib]
 name = "trust_dns_resolver"
 path = "src/lib.rs"
 
 [dependencies]
-cfg-if = "0.1"
+cfg-if = "<=0.1.9"
```
Comment (bluejekyll) [inline crates/resolver/Cargo.toml]: see if this bound can be removed.

File: crates/server/src/server/server_future.rs [hunk 5]
```diff
@@ -57,7 +59,7 @@ impl<T: RequestHandler> ServerFuture<T> {
         let handler = self.handler.clone();
 
         // this spawns a ForEach future which handles all the requests into a Handler.
-        tokio_executor::spawn(
+        let join_handle = runtime.spawn(
```
Comment (bluejekyll) [inline crates/server/src/server/server_future.rs]: file issue to revisit error handling in light of the new JoinHandle

File: crates/server/src/server/server_future.rs [hunk 6]
```diff
@@ -443,33 +473,51 @@ impl<T: RequestHandler> ServerFuture<T> {
...
+                    .await
+            }
         });
 
+        self.joins.push(join);
         Ok(())
     }
+
+    /// This will run until all background tasks of the trust_dns_server end.
+    pub async fn run(self) {
```
Comment (bluejekyll) [inline crates/server/src/server/server_future.rs]: rename to `block_until_done`

**Comment by bluejekyll** [comment]
ok, I believe this is ready to merge.

**Pull request merged**